### PR TITLE
planner: implement a parameterizer for general plan cache based on the current parser

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #36598

Problem Summary: planner: implement a parameterizer for general plan cache based on the current parser

### What is changed and how it works?

planner: implement a parameterizer for general plan cache based on the current parser

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cache_param.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+		n.OrderBy != nil || n.Limit != nil || // order / limit
+		n.LockInfo != nil || n.SelectIntoOpt != nil { // lock info
+		return false
+	}
+	if n.From == nil || n.From.TableRefs == nil { // not from a general table
+		return false
+	}
+	if n.From.TableRefs.Right != nil { // no join
+		return false
+	}
```
Comment (Reminiscent) [inline planner/core/plan_cache_param.go:160]: Maybe should add this part 
```
switch x := n.From.TableRefs.Left.(type) {
 		case *ast.TableSource:
 			_, isTableName := x.Source.(*ast.TableName)
 			if !isTableName {
 				return false
 			}
 		}
```
For example, subquery or something else can also match this part. Please add some test cases for this.

File: planner/core/plan_cache_param.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	return pr.err
+}
+
+// Available4GeneralPlanCache returns whether this AST is available for general plan cache.
+func Available4GeneralPlanCache(sctx sessionctx.Context, stmt ast.StmtNode) bool {
+	// Now only support: select {col} from {single-table} where {cond} and {cond} ...
+	n, ok := stmt.(*ast.SelectStmt)
+	if !ok { // TODO: support DML statements
+		return false
+	}
```
Comment (Reminiscent) [inline planner/core/plan_cache_param.go:139]: Should we check ```
if selectStmt.Kind != ast.SelectStmtKindSelect {
 		return false
 	}
```

File: planner/core/plan_cache_param.go (hunk 1 above)
Comment (qw4990) [reply planner/core/plan_cache_param.go:152]: How about letting us improve it and do more detailed checks on the following several PRs? The main purpose of this PR is to introduce `ParameterizeAST` and `RestoreASTWithParams`.

Comment (qw4990) [reply planner/core/plan_cache_param.go]: Emm, I removed this function. Let's implement this checker on the next PR.

**Review by Reminiscent — APPROVED** [review]

**Review by Yisaer — APPROVED** [review]

**Pull request merged**